# Study on Absenteeism

![title](Features_description.png)

### We are going to try and find possible reasons for absenteeism and which employees are likely to go missing during working hours

In [743]:
import pandas as pd
import numpy as np

In [744]:
raw_data = pd.read_csv('Absenteeism_data.csv')

In [745]:
df = raw_data.copy()

In [746]:
pd.options.display.max_columns = None
display(df)

ID  Reason for Absence        Date  Transportation Expense  \
0    11                  26  07/07/2015                     289   
1    36                   0  14/07/2015                     118   
2     3                  23  15/07/2015                     179   
3     7                   7  16/07/2015                     279   
4    11                  23  23/07/2015                     289   
..   ..                 ...         ...                     ...   
695  17                  10  23/05/2018                     179   
696  28                   6  23/05/2018                     225   
697  18                  10  24/05/2018                     330   
698  25                  23  24/05/2018                     235   
699  15                  28  31/05/2018                     291   

     Distance to Work  Age  Daily Work Load Average  Body Mass Index  \
0                  36   33                  239.554               30   
1                  13   50                  239.554               31   
2                  51   38                  239.554               31   
3                   5   39                  239.554               24   
4                  36   33                  239.554               30   
..                ...  ...                      ...              ...   
695                22   40                  237.656               22   
696                26   28                  237.656               24   
697                16   28                  237.656               25   
698                16   32                  237.656               25   
699                31   40                  237.656               25   

     Education  Children  Pets  Absenteeism Time in Hours  
0            1         2     1                          4  
1            1         1     0                          0  
2            1         0     0                          2  
3            1         2     0                          4  
4            1         2     1                          2  
..         ...       ...   ...                        ...  
695          2         2     0                          8  
696          1         1     2                          3  
697          2         0     0                          8  
698          3         0     0                          2  
699          1         1     1                          2  

[700 rows x 12 columns]

In [747]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ID                         700 non-null    int64  
 1   Reason for Absence         700 non-null    int64  
 2   Date                       700 non-null    object 
 3   Transportation Expense     700 non-null    int64  
 4   Distance to Work           700 non-null    int64  
 5   Age                        700 non-null    int64  
 6   Daily Work Load Average    700 non-null    float64
 7   Body Mass Index            700 non-null    int64  
 8   Education                  700 non-null    int64  
 9   Children                   700 non-null    int64  
 10  Pets                       700 non-null    int64  
 11  Absenteeism Time in Hours  700 non-null    int64  
dtypes: float64(1), int64(10), object(1)
memory usage: 65.8+ KB


Date is an object, not datetime

In [748]:
df.isna().any()

ID                           False
Reason for Absence           False
Date                         False
Transportation Expense       False
Distance to Work             False
Age                          False
Daily Work Load Average      False
Body Mass Index              False
Education                    False
Children                     False
Pets                         False
Absenteeism Time in Hours    False
dtype: bool

let's remember this for now

We don't need the ID column so it's a good idea to drop it

In [749]:
df = df.drop(['ID'], axis=1)
df.head()

Reason for Absence        Date  Transportation Expense  Distance to Work  \
0                  26  07/07/2015                     289                36   
1                   0  14/07/2015                     118                13   
2                  23  15/07/2015                     179                51   
3                   7  16/07/2015                     279                 5   
4                  23  23/07/2015                     289                36   

   Age  Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0   33                  239.554               30          1         2     1   
1   50                  239.554               31          1         1     0   
2   38                  239.554               31          1         0     0   
3   39                  239.554               24          1         2     0   
4   33                  239.554               30          1         2     1   

   Absenteeism Time in Hours  
0                          4  
1                          0  
2                          2  
3                          4  
4                          2

we need to classify the reasons for absenteeism, we don't need this many variables

In [750]:
reason_columns = pd.get_dummies(df['Reason for Absence'])
reason_columns

0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  \
0     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
1     1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3     0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   
4     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..   
695   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   
696   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   
697   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   
698   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
699   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

     18  19  21  22  23  24  25  26  27  28  
0     0   0   0   0   0   0   0   1   0   0  
1     0   0   0   0   0   0   0   0   0   0  
2     0   0   0   0   1   0   0   0   0   0  
3     0   0   0   0   0   0   0   0   0   0  
4     0   0   0   0   1   0   0   0   0   0  
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  
695   0   0   0   0   0   0   0   0   0   0  
696   0   0   0   0   0   0   0   0   0   0  
697   0   0   0   0   0   0   0   0   0   0  
698   0   0   0   0   1   0   0   0   0   0  
699   0   0   0   0   0   0   0   0   0   1  

[700 rows x 28 columns]

In [751]:
reason_columns['check'] = reason_columns.sum(axis=1)
reason_columns

0  1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  21  \
0    0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
1    1  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
2    0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
3    0  0  0  0  0  0  0  1  0  0   0   0   0   0   0   0   0   0   0   0   0   
4    0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
..  .. .. .. .. .. .. .. .. .. ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..   
695  0  0  0  0  0  0  0  0  0  0   1   0   0   0   0   0   0   0   0   0   0   
696  0  0  0  0  0  0  1  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
697  0  0  0  0  0  0  0  0  0  0   1   0   0   0   0   0   0   0   0   0   0   
698  0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
699  0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   

     22  23  24  25  26  27  28  check  
0     0   0   0   0   1   0   0      1  
1     0   0   0   0   0   0   0      1  
2     0   1   0   0   0   0   0      1  
3     0   0   0   0   0   0   0      1  
4     0   1   0   0   0   0   0      1  
..   ..  ..  ..  ..  ..  ..  ..    ...  
695   0   0   0   0   0   0   0      1  
696   0   0   0   0   0   0   0      1  
697   0   0   0   0   0   0   0      1  
698   0   1   0   0   0   0   0      1  
699   0   0   0   0   0   0   1      1  

[700 rows x 29 columns]

In [752]:
reason_columns['check'].unique()

array([1], dtype=int64)

In [753]:
reason_columns['check'].sum()

700

In [754]:
reason_columns = reason_columns.drop(['check'], axis=1)
reason_columns

0  1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  21  \
0    0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
1    1  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
2    0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
3    0  0  0  0  0  0  0  1  0  0   0   0   0   0   0   0   0   0   0   0   0   
4    0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
..  .. .. .. .. .. .. .. .. .. ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..   
695  0  0  0  0  0  0  0  0  0  0   1   0   0   0   0   0   0   0   0   0   0   
696  0  0  0  0  0  0  1  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
697  0  0  0  0  0  0  0  0  0  0   1   0   0   0   0   0   0   0   0   0   0   
698  0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   
699  0  0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0   0   

     22  23  24  25  26  27  28  
0     0   0   0   0   1   0   0  
1     0   0   0   0   0   0   0  
2     0   1   0   0   0   0   0  
3     0   0   0   0   0   0   0  
4     0   1   0   0   0   0   0  
..   ..  ..  ..  ..  ..  ..  ..  
695   0   0   0   0   0   0   0  
696   0   0   0   0   0   0   0  
697   0   0   0   0   0   0   0  
698   0   1   0   0   0   0   0  
699   0   0   0   0   0   0   1  

[700 rows x 28 columns]

In [755]:
reason_columns = pd.get_dummies(df['Reason for Absence'], drop_first=True)
reason_columns

1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  \
0     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
1     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3     0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   
4     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..   
695   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   
696   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   
697   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   
698   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
699   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

     19  21  22  23  24  25  26  27  28  
0     0   0   0   0   0   0   1   0   0  
1     0   0   0   0   0   0   0   0   0  
2     0   0   0   1   0   0   0   0   0  
3     0   0   0   0   0   0   0   0   0  
4     0   0   0   1   0   0   0   0   0  
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  
695   0   0   0   0   0   0   0   0   0  
696   0   0   0   0   0   0   0   0   0  
697   0   0   0   0   0   0   0   0   0  
698   0   0   0   1   0   0   0   0   0  
699   0   0   0   0   0   0   0   0   1  

[700 rows x 27 columns]

looking at the feature descriptions image we now that reasons 1 through 14 are diseases, 15 through 17 are pregnancy related, 18 through 21 are either poison or factors like obesity and 22 through 28 are minor reasons like doctor appointments

we are going to use the max module to keep 1 when this new dataframe has 1 in any of these columns

In [756]:
reason_1 = reason_columns.loc[:, 1:14].max(axis=1)
reason_1

0      0
1      0
2      0
3      1
4      0
      ..
695    1
696    1
697    1
698    0
699    0
Length: 700, dtype: uint8

In [757]:
reason_2 = reason_columns.loc[:, 15:17].max(axis=1)
reason_2

0      0
1      0
2      0
3      0
4      0
      ..
695    0
696    0
697    0
698    0
699    0
Length: 700, dtype: uint8

In [758]:
reason_3 = reason_columns.loc[:, 18:21].max(axis=1)
reason_3

0      0
1      0
2      0
3      0
4      0
      ..
695    0
696    0
697    0
698    0
699    0
Length: 700, dtype: uint8

In [759]:
reason_4 = reason_columns.loc[:, 22:].max(axis=1)
reason_4

0      1
1      0
2      1
3      0
4      1
      ..
695    0
696    0
697    0
698    1
699    1
Length: 700, dtype: uint8

In [760]:
df = df.drop(['Reason for Absence'], axis=1)
df

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours  
0                            4  
1                            0  
2                            2  
3                            4  
4                            2  
..                         ...  
695                          8  
696                          3  
697                          8  
698                          2  
699                          2  

[700 rows x 10 columns]

And now we insert them into the dataframe

In [761]:
df.insert(0, 'Reason 1', reason_1)
df.insert(1, 'Reason 2', reason_2)
df.insert(2, 'Reason 3', reason_3)
df.insert(3, 'Reason 4', reason_4)
df

Reason 1  Reason 2  Reason 3  Reason 4        Date  \
0           0         0         0         1  07/07/2015   
1           0         0         0         0  14/07/2015   
2           0         0         0         1  15/07/2015   
3           1         0         0         0  16/07/2015   
4           0         0         0         1  23/07/2015   
..        ...       ...       ...       ...         ...   
695         1         0         0         0  23/05/2018   
696         1         0         0         0  23/05/2018   
697         1         0         0         0  24/05/2018   
698         0         0         0         1  24/05/2018   
699         0         0         0         1  31/05/2018   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0                 30          1         2     1                          4  
1                 31          1         1     0                          0  
2                 31          1         0     0                          2  
3                 24          1         2     0                          4  
4                 30          1         2     1                          2  
..               ...        ...       ...   ...                        ...  
695               22          2         2     0                          8  
696               24          1         1     2                          3  
697               25          2         0     0                          8  
698               25          3         0     0                          2  
699               25          1         1     1                          2  

[700 rows x 14 columns]

In [762]:
df_reasons = df.copy()

In [763]:
df_reasons.Date = pd.to_datetime(df_reasons.Date, format= '%d/%m/%Y')

In [764]:
df_reasons.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   Reason 1                   700 non-null    uint8         
 1   Reason 2                   700 non-null    uint8         
 2   Reason 3                   700 non-null    uint8         
 3   Reason 4                   700 non-null    uint8         
 4   Date                       700 non-null    datetime64[ns]
 5   Transportation Expense     700 non-null    int64         
 6   Distance to Work           700 non-null    int64         
 7   Age                        700 non-null    int64         
 8   Daily Work Load Average    700 non-null    float64       
 9   Body Mass Index            700 non-null    int64         
 10  Education                  700 non-null    int64         
 11  Children                   700 non-null    int64         
 12  Pets    

Date is now datetime

In [765]:
df_reasons.head()

Reason 1  Reason 2  Reason 3  Reason 4       Date  Transportation Expense  \
0         0         0         0         1 2015-07-07                     289   
1         0         0         0         0 2015-07-14                     118   
2         0         0         0         1 2015-07-15                     179   
3         1         0         0         0 2015-07-16                     279   
4         0         0         0         1 2015-07-23                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   
3                 5   39                  239.554               24          1   
4                36   33                  239.554               30          1   

   Children  Pets  Absenteeism Time in Hours  
0         2     1                          4  
1         1     0                          0  
2         0     0                          2  
3         2     0                          4  
4         2     1                          2

let's make columns for the day of the week and the month, i add 1 to the weekday value so it goes from 1 to 7 instead of 0 to 6, it makes the analysis more intuitive

In [766]:
months = [df_reasons.Date[i].month for i in range(df_reasons.Date.count())]
day_of_the_week = [df_reasons.Date[i].weekday()+1 for i in range(df_reasons.Date.count())]

and we insert them into the dataframe

In [767]:
df_reasons.insert(5, 'Day of the Week', day_of_the_week)
df_reasons.insert(6, 'Month', months)
df_reasons.head()

Reason 1  Reason 2  Reason 3  Reason 4       Date  Day of the Week  Month  \
0         0         0         0         1 2015-07-07                2      7   
1         0         0         0         0 2015-07-14                2      7   
2         0         0         0         1 2015-07-15                3      7   
3         1         0         0         0 2015-07-16                4      7   
4         0         0         0         1 2015-07-23                4      7   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          1         2     1                          4  
1               31          1         1     0                          0  
2               31          1         0     0                          2  
3               24          1         2     0                          4  
4               30          1         2     1                          2

seems fine

In [768]:
sorted(df_reasons['Day of the Week'].unique())

[1, 2, 3, 4, 5, 6, 7]

we can drop the date column since the year doesn't provide any value

In [769]:
df_reasons = df_reasons.drop(['Date'], axis=1)
df_reasons.head()

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0         0         0         0         1                2      7   
1         0         0         0         0                2      7   
2         0         0         0         1                3      7   
3         1         0         0         0                4      7   
4         0         0         0         1                4      7   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          1         2     1                          4  
1               31          1         1     0                          0  
2               31          1         0     0                          2  
3               24          1         2     0                          4  
4               30          1         2     1                          2

In [770]:
df_reasons.Education.unique()

array([1, 3, 2, 4], dtype=int64)

In [771]:
df_reasons.Education.value_counts()

1    583
3     73
2     40
4      4
Name: Education, dtype: int64

Here we can that 583 possess the lowest level of education and 117 higher so we can divide this in 2 groups, low education and high education

In [772]:
df_reasons['Education'] = df_reasons['Education'].map({1:0, 2:1, 3:1, 4:1})
df_reasons['Education'].unique()

array([0, 1], dtype=int64)

In [773]:
df_reasons['Education'].value_counts()

0    583
1    117
Name: Education, dtype: int64

Here we will do the same with body weight, pets and children

In [774]:
df_reasons.Pets.unique()

array([1, 0, 4, 2, 5, 8], dtype=int64)

0 if they don't have pets and 1 if they do

In [775]:
df_reasons.Pets = df_reasons.Pets.map({1:1, 0:0, 4:1, 2:1, 5:1, 8:1})
df_reasons.Pets.unique()

array([1, 0], dtype=int64)

In [776]:
df_reasons.Children.unique()

array([2, 1, 0, 4, 3], dtype=int64)

In [777]:
df_reasons.Children = df_reasons.Children.map({2:1, 1:1, 0:0, 4:1, 3:1})
df_reasons.Children.unique()

array([1, 0], dtype=int64)

##### BMI Categories:
* Underweight = <18.5
* Normal weight = 18.5–24.9
* Overweight = 25–29.9
* Obesity = BMI of 30 or greater

we already know the dataframe has 700 rows so i'm gonna use that from now on

In [778]:
underweight = []
normal_weight = []
overweight = []
obese = []

for i in range(len(df_reasons['Body Mass Index'])):
    if df_reasons['Body Mass Index'][i] < 18.5:
        underweight.append(1)
    else:
        underweight.append(0)

for i in range(len(df_reasons['Body Mass Index'])):
    if df_reasons['Body Mass Index'][i] >= 18.5 and df_reasons['Body Mass Index'][i] <= 24.9:
        normal_weight.append(1)
    else:
        normal_weight.append(0)

for i in range(len(df_reasons['Body Mass Index'])):
    if df_reasons['Body Mass Index'][i] > 24.9 and df_reasons['Body Mass Index'][i] <= 29.9:
        overweight.append(1)
    else:
        overweight.append(0)

for i in range(len(df_reasons['Body Mass Index'])):
    if df_reasons['Body Mass Index'][i] > 29.9:
        obese.append(1)
    else:
        obese.append(0)

In [779]:
print(set(underweight)), print(len(underweight))
print(set(normal_weight)), print(len(normal_weight))
print(set(overweight)), print(len(overweight))
print(set(obese)), print(len(obese))

{0}
700
{0, 1}
700
{0, 1}
700
{0, 1}
700


(None, None)

no underweight people, that's good isn't it, that means we won't be using that list

In [780]:
df_reasons.head()

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0         0         0         0         1                2      7   
1         0         0         0         0                2      7   
2         0         0         0         1                3      7   
3         1         0         0         0                4      7   
4         0         0         0         1                4      7   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         1     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         1     0                          4  
4               30          0         1     1                          2

let's add the weight columns

In [781]:
df_reasons.insert(10, 'Normal Weight', normal_weight)
df_reasons.insert(11, 'Overweight', overweight)
df_reasons.insert(12,'Obese', obese)
df_reasons.head()

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0         0         0         0         1                2      7   
1         0         0         0         0                2      7   
2         0         0         0         1                3      7   
3         1         0         0         0                4      7   
4         0         0         0         1                4      7   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Normal Weight  Overweight  Obese  Body Mass Index  Education  Children  \
0              0           0      1               30          0         1   
1              0           0      1               31          0         1   
2              0           0      1               31          0         0   
3              1           0      0               24          0         1   
4              0           0      1               30          0         1   

   Pets  Absenteeism Time in Hours  
0     1                          4  
1     0                          0  
2     0                          2  
3     0                          4  
4     1                          2

now we can drop the body mass index column

In [782]:
df_reasons = df_reasons.drop(['Body Mass Index'], axis=1)
df_reasons.head()

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0         0         0         0         1                2      7   
1         0         0         0         0                2      7   
2         0         0         0         1                3      7   
3         1         0         0         0                4      7   
4         0         0         0         1                4      7   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Normal Weight  Overweight  Obese  Education  Children  Pets  \
0              0           0      1          0         1     1   
1              0           0      1          0         1     0   
2              0           0      1          0         0     0   
3              1           0      0          0         1     0   
4              0           0      1          0         1     1   

   Absenteeism Time in Hours  
0                          4  
1                          0  
2                          2  
3                          4  
4                          2

In [783]:
df.iloc[df['Absenteeism Time in Hours'].idxmax()]

Reason 1                              1
Reason 2                              0
Reason 3                              0
Reason 4                              0
Date                         14/11/2016
Transportation Expense              155
Distance to Work                     12
Age                                  34
Daily Work Load Average         284.031
Body Mass Index                      25
Education                             1
Children                              2
Pets                                  0
Absenteeism Time in Hours           120
Name: 323, dtype: object

Checking for weird numbers, we can see someone dissapeared for 120 hours, that's weird.

Looking for outliers using the 3 sigma rule

In [784]:
absenteeism_mean = df_reasons[['Absenteeism Time in Hours']].mean()[0]
absenteeism_mean

6.761428571428572

the average numbers of hours missing for the employees is 6 to 7

In [785]:
absenteeism_std = df_reasons[['Absenteeism Time in Hours']].std()[0]
absenteeism_std

12.670082283897274

the standard deviation is waaaay off, huh

In [786]:
absenteeism_3sigma = absenteeism_mean+(absenteeism_std*3)
absenteeism_3sigma

44.7716754231204

just to be sure let's add another standard deviation to the 3 sigma

In [787]:
absenteeism_3sigma = absenteeism_3sigma+absenteeism_std
absenteeism_3sigma

57.44175770701767

that's generous enough

In [788]:
df_outliers = df_reasons[df['Absenteeism Time in Hours'] > absenteeism_3sigma]
df_outliers

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
99          0         0         1         0                2     12   
231         1         0         0         0                3      6   
299         1         0         0         0                5     10   
323         1         0         0         0                1     11   
420         1         0         0         0                3      4   
471         1         0         0         0                2      7   
569         0         0         1         0                2     12   
622         1         0         0         0                2      3   
652         0         0         1         0                1      3   
682         0         0         1         0                5      5   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
99                      118                10   37                  261.306   
231                     118                13   50                  377.550   
299                     179                26   30                  265.017   
323                     155                12   34                  284.031   
420                     118                13   50                  239.409   
471                     225                26   28                  230.290   
569                     155                12   34                  280.549   
622                     228                14   58                  222.196   
652                     289                36   33                  222.196   
682                     300                26   43                  237.656   

     Normal Weight  Overweight  Obese  Education  Children  Pets  \
99               0           1      0          0         0     0   
231              0           0      1          0         1     0   
299              1           0      0          1         0     0   
323              0           1      0          0         1     0   
420              0           0      1          0         1     0   
471              1           0      0          0         1     1   
569              0           1      0          0         1     0   
622              1           0      0          0         1     1   
652              0           0      1          0         1     1   
682              0           1      0          0         1     1   

     Absenteeism Time in Hours  
99                          64  
231                         80  
299                         64  
323                        120  
420                        120  
471                        112  
569                         80  
622                        112  
652                        104  
682                         64

that' interesting, almost all employees who have been absent for extended periods of time have children, nothing really stands out other than that

i'm not going to do anything with these, though, nothing really stands out and the absent hours aren't particularly ridiculous, i just wanted to see if i had crazy outliers

according to google it is considered excess absenteeism when the person is absent for more that 3 hours every 90 days or 12 hours in a year, the dataset shows registers from 2015 to 2018 so 3 years. Anyone with more than 36 hours absent is gonna be considered an excessive absentee

first let's get the IDs

In [789]:
ID_list = list(raw_data.ID.unique())

then we initialize the dictionary that will save the hours absent

In [790]:
absent_hours = {i:0 for i in ID_list}

add the hours to each employee

In [791]:
for i in range(700):
    absent_hours[raw_data.iloc[i]['ID']] = absent_hours[raw_data.iloc[i]['ID']]+raw_data.iloc[i]['Absenteeism Time in Hours']

looks good

In [792]:
absent_hours

{11: 442,
 36: 284,
 3: 482,
 7: 30,
 10: 170,
 20: 306,
 14: 466,
 1: 117,
 24: 254,
 6: 56,
 33: 73,
 18: 118,
 30: 31,
 2: 25,
 19: 6,
 27: 27,
 34: 314,
 5: 100,
 15: 251,
 29: 21,
 28: 338,
 13: 103,
 22: 241,
 17: 118,
 31: 16,
 23: 32,
 32: 16,
 9: 134,
 26: 83,
 21: 16,
 8: 0,
 25: 42,
 12: 13,
 16: 8}

now we label the excessive absentees as 1 and the rest as 0

In [793]:
for i, (key, value) in enumerate(absent_hours.items()):
    if value > 36:
        absent_hours[key] = 1
    else:
        absent_hours[key] = 0
    

looks good

In [794]:
absent_hours

{11: 1,
 36: 1,
 3: 1,
 7: 0,
 10: 1,
 20: 1,
 14: 1,
 1: 1,
 24: 1,
 6: 1,
 33: 1,
 18: 1,
 30: 0,
 2: 0,
 19: 0,
 27: 0,
 34: 1,
 5: 1,
 15: 1,
 29: 0,
 28: 1,
 13: 1,
 22: 1,
 17: 1,
 31: 0,
 23: 0,
 32: 0,
 9: 1,
 26: 1,
 21: 0,
 8: 0,
 25: 1,
 12: 0,
 16: 0}

now we can create the excessive absenteeism column

In [795]:
excessive_absenteeism = []
for i in range(700):
    if absent_hours[raw_data.iloc[i]['ID']] == 1:
        excessive_absenteeism.append(1)
    else:
        excessive_absenteeism.append(0)

looks good

In [796]:
len(excessive_absenteeism)

700

that's not good, it's terribly skewed, it's 3 hours every 3 months so one hour a month i guess?

In [797]:
df_reasons.loc[df_reasons['Absenteeism Time in Hours']>1]

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0           0         0         0         1                2      7   
2           0         0         0         1                3      7   
3           1         0         0         0                4      7   
4           0         0         0         1                4      7   
5           0         0         0         1                5      7   
..        ...       ...       ...       ...              ...    ...   
695         1         0         0         0                3      5   
696         1         0         0         0                3      5   
697         1         0         0         0                4      5   
698         0         0         0         1                4      5   
699         0         0         0         1                4      5   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
5                       179                51   38                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Normal Weight  Overweight  Obese  Education  Children  Pets  \
0                0           0      1          0         1     1   
2                0           0      1          0         0     0   
3                1           0      0          0         1     0   
4                0           0      1          0         1     1   
5                0           0      1          0         0     0   
..             ...         ...    ...        ...       ...   ...   
695              1           0      0          1         1     0   
696              1           0      0          0         1     1   
697              0           1      0          1         0     0   
698              0           1      0          1         0     0   
699              0           1      0          0         1     1   

     Absenteeism Time in Hours  
0                            4  
2                            2  
3                            4  
4                            2  
5                            2  
..                         ...  
695                          8  
696                          3  
697                          8  
698                          2  
699                          2  

[574 rows x 17 columns]

how about 2 hours

In [798]:
df_reasons.loc[df_reasons['Absenteeism Time in Hours']>2]

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0           0         0         0         1                2      7   
3           1         0         0         0                4      7   
6           0         0         0         1                5      7   
7           0         0         0         1                5      7   
8           0         0         1         0                1      7   
..        ...       ...       ...       ...              ...    ...   
693         1         0         0         0                1      5   
694         0         0         0         1                3      5   
695         1         0         0         0                3      5   
696         1         0         0         0                3      5   
697         1         0         0         0                4      5   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
3                       279                 5   39                  239.554   
6                       361                52   28                  239.554   
7                       260                50   36                  239.554   
8                       155                12   34                  239.554   
..                      ...               ...  ...                      ...   
693                     235                16   32                  237.656   
694                     291                31   40                  237.656   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   

     Normal Weight  Overweight  Obese  Education  Children  Pets  \
0                0           0      1          0         1     1   
3                1           0      0          0         1     0   
6                0           1      0          0         1     1   
7                1           0      0          0         1     0   
8                0           1      0          0         1     0   
..             ...         ...    ...        ...       ...   ...   
693              0           1      0          1         0     0   
694              0           1      0          0         1     1   
695              1           0      0          1         1     0   
696              1           0      0          0         1     1   
697              0           1      0          1         0     0   

     Absenteeism Time in Hours  
0                            4  
3                            4  
6                            8  
7                            4  
8                           40  
..                         ...  
693                          8  
694                          8  
695                          8  
696                          3  
697                          8  

[425 rows x 17 columns]

3?

In [799]:
df_reasons.loc[df_reasons['Absenteeism Time in Hours']>3]

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0           0         0         0         1                2      7   
3           1         0         0         0                4      7   
6           0         0         0         1                5      7   
7           0         0         0         1                5      7   
8           0         0         1         0                1      7   
..        ...       ...       ...       ...              ...    ...   
692         1         0         0         0                1      5   
693         1         0         0         0                1      5   
694         0         0         0         1                3      5   
695         1         0         0         0                3      5   
697         1         0         0         0                4      5   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
3                       279                 5   39                  239.554   
6                       361                52   28                  239.554   
7                       260                50   36                  239.554   
8                       155                12   34                  239.554   
..                      ...               ...  ...                      ...   
692                     155                12   34                  237.656   
693                     235                16   32                  237.656   
694                     291                31   40                  237.656   
695                     179                22   40                  237.656   
697                     330                16   28                  237.656   

     Normal Weight  Overweight  Obese  Education  Children  Pets  \
0                0           0      1          0         1     1   
3                1           0      0          0         1     0   
6                0           1      0          0         1     1   
7                1           0      0          0         1     0   
8                0           1      0          0         1     0   
..             ...         ...    ...        ...       ...   ...   
692              0           1      0          0         1     0   
693              0           1      0          1         0     0   
694              0           1      0          0         1     1   
695              1           0      0          1         1     0   
697              0           1      0          1         0     0   

     Absenteeism Time in Hours  
0                            4  
3                            4  
6                            8  
7                            4  
8                           40  
..                         ...  
692                         48  
693                          8  
694                          8  
695                          8  
697                          8  

[319 rows x 17 columns]

sounds good, this is a weird dataset

let's map the absence column with 1s for employees with more than 3 hours absent in a register

In [800]:
excess_absence = np.where(df_reasons['Absenteeism Time in Hours']>3,1,0)
excess_absence

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [801]:
df_reasons.insert(17, 'Excessive Absenteeism', excess_absence)
df_reasons

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0           0         0         0         1                2      7   
1           0         0         0         0                2      7   
2           0         0         0         1                3      7   
3           1         0         0         0                4      7   
4           0         0         0         1                4      7   
..        ...       ...       ...       ...              ...    ...   
695         1         0         0         0                3      5   
696         1         0         0         0                3      5   
697         1         0         0         0                4      5   
698         0         0         0         1                4      5   
699         0         0         0         1                4      5   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Normal Weight  Overweight  Obese  Education  Children  Pets  \
0                0           0      1          0         1     1   
1                0           0      1          0         1     0   
2                0           0      1          0         0     0   
3                1           0      0          0         1     0   
4                0           0      1          0         1     1   
..             ...         ...    ...        ...       ...   ...   
695              1           0      0          1         1     0   
696              1           0      0          0         1     1   
697              0           1      0          1         0     0   
698              0           1      0          1         0     0   
699              0           1      0          0         1     1   

     Absenteeism Time in Hours  Excessive Absenteeism  
0                            4                      1  
1                            0                      0  
2                            2                      0  
3                            4                      1  
4                            2                      0  
..                         ...                    ...  
695                          8                      1  
696                          3                      0  
697                          8                      1  
698                          2                      0  
699                          2                      0  

[700 rows x 18 columns]

In [802]:
df_reasons.drop(['Absenteeism Time in Hours'], axis=1, inplace=True)
df_reasons

Reason 1  Reason 2  Reason 3  Reason 4  Day of the Week  Month  \
0           0         0         0         1                2      7   
1           0         0         0         0                2      7   
2           0         0         0         1                3      7   
3           1         0         0         0                4      7   
4           0         0         0         1                4      7   
..        ...       ...       ...       ...              ...    ...   
695         1         0         0         0                3      5   
696         1         0         0         0                3      5   
697         1         0         0         0                4      5   
698         0         0         0         1                4      5   
699         0         0         0         1                4      5   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Normal Weight  Overweight  Obese  Education  Children  Pets  \
0                0           0      1          0         1     1   
1                0           0      1          0         1     0   
2                0           0      1          0         0     0   
3                1           0      0          0         1     0   
4                0           0      1          0         1     1   
..             ...         ...    ...        ...       ...   ...   
695              1           0      0          1         1     0   
696              1           0      0          0         1     1   
697              0           1      0          1         0     0   
698              0           1      0          1         0     0   
699              0           1      0          0         1     1   

     Excessive Absenteeism  
0                        1  
1                        0  
2                        0  
3                        1  
4                        0  
..                     ...  
695                      1  
696                      0  
697                      1  
698                      0  
699                      0  

[700 rows x 17 columns]

This dataset looks clean enough, now we can get started with the machine learning model

In [644]:
df_reasons.to_csv('Absenteeism_preprocessed.csv', index=False)